File: frontend/src/types/analysis.ts
/**
 * Analysis Progress Types
 * Defines types for SSE events and progress tracking during analysis
 */

export const AnalysisStatus = {
  UPLOADING: 'uploading',
  ANALYZING_RESUME: 'analyzing_resume',
  ANALYZING_JD: 'analyzing_jd',
  EXTRACTING: 'extracting',
  SCORING: 'scoring',
  ASSESSING_JOB_FIT: 'assessing_job_fit',
  GENERATING_RECOMMENDATIONS: 'generating_recommendations',
  SAVING: 'saving',
  COMPLETE: 'complete',
  ERROR: 'error',
} as const;

export type AnalysisStatusType = typeof AnalysisStatus[keyof typeof AnalysisStatus];

export interface AnalysisProgressEvent {
  status: AnalysisStatusType | string;
  message: string;
  progress: number; // 0-100
  stage: string;
  data?: any; // Final analysis results on complete
}

export interface AnalysisStage {
  id: string;
  status: AnalysisStatusType | string;
  message: string;
  progress: number;
  completed: boolean;
  icon: string; // emoji or icon name
}

export interface AnalysisProgressState {
  currentStatus: AnalysisStatusType | string;
  progress: number;
  message: string;
  stages: AnalysisStage[];
  isComplete: boolean;
  hasError: boolean;
  error?: string;
  analysisResult?: any;
  resumeDbId?: string | null;
  jdDbId?: string | null;
}

export interface LearningResource {
  title: string;
  description: string;
  category: string; // "technical_skills" | "soft_skills" | "certifications"
  resource_type: string; // "course" | "certification" | "tutorial" | "book" | "video"
  url: string;
  estimated_hours: number;
}

export interface CategoryScores {
  skills_match?: number;
  experience_alignment?: number;
  education_and_certifications?: number;
  achievements_and_outcomes?: number;
  soft_skills_and_culture?: number;
  [key: string]: number | undefined;
}

export interface AnalysisResult {
  analysis_id: string;
  overall_score: number;
  fit_classification: 'GOOD_FIT' | 'PARTIAL_FIT' | 'NOT_FIT';
  fit_rationale: string;
  category_scores: CategoryScores;
  recommendations: string[]; // Array of actionable recommendations in second-person voice
  learning_resources: LearningResource[]; // Array of curated resources with real URLs
  progress: number;
  message: string;
}

export interface AnalysisResponse {
  status: string;
  message: string;
  progress: number;
  stage: string;
  data?: AnalysisResult;
}

File: backend/api/types/analysis.py
"""
Data models for analysis requests and responses.
"""

from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Literal
from datetime import datetime
from enum import Enum
from fastapi import Body, Depends

from api._utils.auth import get_current_user_id


class AnalyzeRequestParams:
    """
    Dependency class for analyze request parameters.
    
    Extracts resume_db_id, jd_db_id from request body and user_id from JWT token.
    
    ### Usage:
    
    ```python
    @router.post("/analyze")
    async def analyze_endpoint(params: AnalyzeRequestParams = Depends()):
        resume_db_id = params.resume_db_id
        jd_db_id = params.jd_db_id
        user_id = params.user_id
    ```
    """
    
    def __init__(
        self,
        resume_db_id: str = Body(..., description="UUID of resume record from Supabase"),
        jd_db_id: str = Body(..., description="UUID of JD record from Supabase"),
        user_id: str = Depends(get_current_user_id),
    ):
        """
        Initialize analyze request parameters.
        
        Args:
            resume_db_id: UUID of resume stored in Supabase
            jd_db_id: UUID of job description stored in Supabase
            user_id: Authenticated user ID (extracted from JWT token)
        """
        self.resume_db_id = resume_db_id
        self.jd_db_id = jd_db_id
        self.user_id = user_id


class FitClassification(str, Enum):
    """Enum for fit classifications."""

    GOOD_FIT = "GOOD_FIT"
    PARTIAL_FIT = "PARTIAL_FIT"
    NOT_FIT = "NOT_FIT"


class CategoryScore(BaseModel):
    """Score for a single analysis category."""

    category: str
    score: float = Field(..., ge=0, le=100)
    reason: str
    weight: float = Field(..., ge=0, le=1)


class LearningResource(BaseModel):
    """Recommended learning resource."""

    title: str
    url: str
    type: Literal["course", "certification", "article", "book"]
    relevance: str


class AnalysisRequest(BaseModel):
    """Request to analyze resume against JD."""

    resume_id: str = Field(..., description="UUID of resume")
    jd_id: str = Field(..., description="UUID of job description")


class AnalysisResponse(BaseModel):
    """Complete analysis response."""

    analysis_id: str
    resume_id: str
    jd_id: str

    # Extraction results
    resume_extracted: Dict
    jd_extracted: Dict

    # Category analysis
    category_weights: Dict[str, float]
    category_scores: List[CategoryScore]

    # Overall assessment
    composite_score: float = Field(..., ge=0, le=100)
    fit_classification: FitClassification
    fit_rationale: str

    # Recommendations
    actionable_recommendations: List[str]
    learning_resources: List[LearningResource]

    # Metadata
    analyzed_at: datetime = Field(default_factory=datetime.utcnow)
    processing_time_seconds: float


class AnalysisStatus(BaseModel):
    """Real-time status update during analysis."""

    analysis_id: str
    stage: Literal[
        "starting_analysis",
        "calculating_match_score",
        "assessing_job_fit",
        "generating_recommendations",
        "curating_learning_resources",
        "complete",
        "error",
    ]
    progress: float = Field(..., ge=0, le=100)
    message: str


class CategoryScoreExtended(BaseModel):
    """Extended score with strengths and gaps."""

    category: str
    score: float = Field(..., ge=0, le=100)
    reason: str
    strengths: List[str] = []
    gaps: List[str] = []
    weight: float = Field(..., ge=0, le=1)
